File: insertion_sort.c
#include <stdio.h>

int main()
{
    #define SIZE 21
    int A[SIZE] = {2,165,4,9,23,1,493,5,65,49,4,6,96,2,6,72,2,3,35,0,2};
    
    print_arr(&A,SIZE);
    
    i_sort(&A,SIZE);
    
    print_arr(&A,SIZE);

    return 0;
}

void print_arr(int arr[],int siz) {
    //int siz = (sizeof(arr)/sizeof(arr[0]));
    //int siz = sizeof(arr);
    printf("array size=%d\n",siz);
    
    for(int i=0; i<siz;i++) {
        printf("%d",arr[i]);
        if((i+1)<siz)
            printf(", ");  //dont print after last number
        else
            printf("\n");
    }
}

void i_sort(int arr[], int siz) {
    for(int i=1;i<siz;i++) {
        int j = i;
        while (j > 0 && arr[j-1] > arr[j]) {
            int k = arr[j];
            arr[j] = arr[j-1];
            arr[j-1] = k;
            j = j-1;
        }
    }
}

File: insertion_sort.go
package main
import "fmt"

func main() {
    arr:= []int{2,165,4,9,23,1,493,5,65,49,4,6,96,2,6,72,2,3,35,0,2} 
    
    print_arr(arr)
    
    i_sort(arr)
    
    print_arr(arr)
}

func print_arr(a []int) {
    var siz int = len(a)
    
    for i := 0; i<siz; i++ {
        fmt.Print(a[i])
        if (i+1) < siz {
            fmt.Print(", ")
        } else {
            fmt.Print("\n")
        }
    }

}

func i_sort(a []int) {
    for i:= 1; i<len(a); i++ {
        var j int = i
        for (j > 0) && (a[j-1] > a[j]) {
            var k int = a[j]
            a[j] = a[j-1]
            a[j-1] = k
            j = j-1
        }
    }
}
